### PR TITLE
Fix VCS dependencies

### PR DESCRIPTION
**Pull request by nateprewitt**
~This is **DO NOT MERGE** until at least later today. Still need to find some time to clean up but want to allow some time for review from anyone who's interested.~

This is an attempt to address #204, #229 and #240 which all have the same underlying issue. When dealing with Pipfile and Pipfile.lock entries we've had a lot of issues with casing and name formatting in comparisons. We'd fixed this previously for package names specified by us, or pipenv users, but missed the case with mis-cased names in other packages' `install_requires` lists.

In order to prevent this from occurring in the future, I've transitioned to using a normalized naming scheme derived from pep426 where all names will be lower-cased and use dashes in place of underscores. This should allow uniform comparison and then we will rewrite the file after passing it through `proper_case` on the way out to ensure "pretty" package names.

This fixes part of the VCS problem but exposed another. We currently require hashes for ALL dependencies which pip can't support for packages that aren't hosted on PyPI. Due to this requirement, we will now split vcs dependencies off from PyPI packages before install to keep things working correctly.

This still has a few cosmetic/organization changes that need to be made, along with some testing for the new functions, but this "works" right now. If anyone has comments or wants to test this out you can clone the repository, run `git fetch origin pull/242/head:fix_vcs_dependencies`, and checkout the branch `fix_vcs_dependencies`.

### CODE REVIEW

**Comment by simnalamburt** [comment]
Thank you for making a big change! You did something which others could not. I have tested this branch with the command below, and I would like to make a proposal.

```bash
pip install --upgrade git+https://github.com/nateprewitt/pipenv@fix_vcs_dependencies
```

#### A. Let's make an error when users did not give the `#egg=` parameter.
The right way to install VCS dependencies is this:

```bash
pipenv install git+https://github.com/kennethreitz/requests.git#egg=requests

# [packages.requests]
# git = "https://github.com/kennethreitz/requests.git"
```

But users are able to forget this and use it like this:

```bash
pipenv install git+https://github.com/kennethreitz/requests.git

# [packages.none]
# git = "https://github.com/kennethreitz/requests.git"
```

With current implementation, pipenv just succeeds and installs the dependency as "none". I think this may cause confusion to some users. How do you think about fast-failing this and giving some error message to the users?

It'll be best if we can parse the setup.py and automatically decide the package name, but I have no idea how to implement it.

**Comment by nateprewitt** [comment]
Great catch @simnalamburt, I'll look at getting that added in the morning. Thanks!

**Comment by simnalamburt** [comment]
`crayons` module have never been imported into utils.py, but it was used. Please merge https://github.com/simnalamburt/pipenv/commit/2e840676 this commit before accepting the PR

**Comment by nateprewitt** [comment]
Yep, that was definitely an oversight. Thanks @simnalamburt! Instead of importing crayons, I think we'll just pull the coloring use off the exception message because it's no longer surfaced in the cli.

File: pipenv/utils.py [hunk 1]
```diff
@@ -144,3 +153,35 @@ def is_required_version(version, specified_version):
...
+        def handle_data(self, data):
+            # Remove extra blank data from https://pypi.org/simple
+            data = data.strip()
+            if len(data) > 2:
+                collected.append(data)
+
+    # Hit the simple API.
+    r = requests.get('https://pypi.org/simple/{0}'.format(package_name))
+    if not r.ok:
+        raise IOError('Unable to find package {0} in PyPI repository.'.format(crayons.green(package_name)))
```
Comment (nateprewitt) [inline pipenv/utils.py]: There's no longer a reason to import crayons as of 47b5ded :)

**Comment by simnalamburt** [comment]
Looks good to me now!

**Comment by nateprewitt** [comment]
Alright, I'm planning on leaving this up for a few more days for @kennethreitz to look over if he feels like it. After that I'll merge this in either Friday or Monday and try to push a 3.5.0 release shortly after.

**Comment by simnalamburt** [comment]
No problem! 👍

**Comment by kennethreitz** [comment]
:sparkles: :cake: :sparkles:

**Pull request merged**